cli/src/commands/bookmark/list.rs

```diff
@@ -270,3 +291,470 @@ struct RefListItem {
     /// Remote bookmarks tracked by the primary (or local) bookmark.
     tracked: Vec<Rc<CommitRef>>,
 }
+
+/// Sort key for the `--sort` argument option.
+#[derive(Copy, Clone, PartialEq, Debug, ValueEnum)]
+enum SortKey {
+    Name,
+    #[value(name = "name-")]
+    NameDesc,
+    AuthorName,
+    #[value(name = "author-name-")]
+    AuthorNameDesc,
+    AuthorEmail,
+    #[value(name = "author-email-")]
+    AuthorEmailDesc,
+    AuthorDate,
+    #[value(name = "author-date-")]
+    AuthorDateDesc,
+    CommitterName,
+    #[value(name = "committer-name-")]
+    CommitterNameDesc,
+    CommitterEmail,
+    #[value(name = "committer-email-")]
+    CommitterEmailDesc,
+    CommitterDate,
+    #[value(name = "committer-date-")]
+    CommitterDateDesc,
+}
+
+fn commit_for_ref_list_item(
+    item: &RefListItem,
+    f: impl Fn(CommitId) -> Option<Arc<backend::Commit>>,
+) -> Option<Arc<backend::Commit>> {
+    item.primary
+        .target()
+        .added_ids()
+        .next()
+        .cloned()
+        .and_then(f)
+}
+
+fn sort(
+    bookmark_items: &mut [RefListItem],
+    sort_keys: &[SortKey],
+    f: impl Fn(CommitId) -> Option<Arc<backend::Commit>>,
+) {
+    // Multi-pass sorting, the first key is most significant.
+    for (i, sort_key) in sort_keys.iter().rev().enumerate() {
+        let f = &f;
+        match sort_key {
+            SortKey::Name => {
+                // Bookmarks are already sorted by name.
+                let is_noop = i == 0 && sort_key == &SortKey::Name;
+                if !is_noop {
+                    bookmark_items.sort_by_key(|item| item.primary.name().to_owned());
+                }
+            }
+            SortKey::NameDesc => {
+                bookmark_items.sort_by_key(|item| cmp::Reverse(item.primary.name().to_owned()));
```

Comment:
nit: we'll need to sort by `(primary.name(), primary.remote_name())` to sort untracked remote bookmarks of the same name.